### ISSUE TITLE
Track notification interaction

### ISSUE THREAD

**Issue by galenriley** [Chattahooligans/hooligan-hymnal-app#2]
Which notification brought the user into the app? Let's keep track of that somehow, maybe it'll be useful for analytics later

**Comment by galenriley** [comment]
We want to give admins info about how users are engaging with push notifications. 

The suggested approach is something like this, with no confidence in the API name, collection name, or property names
```
{
  post id,
  count of engagements,
  first engagement timestamp (firstTappedAt),
  most recent timestamp,
  engagements:[
    {
      pushToken,
      engagement timestamp (tappedAt)
    }
  ]
}
```
The API will accept post id, push token, timestamp* 

*Do we want to track local time on the user's device or skip this parameter and use server time?
We have a lot of interest in network congestion in stadium environments, so maybe we timestamp the tap on the client AND when it was received at the server in the array

**Comment by rsinger** [comment]
This look like two distinct models to me: `engagement` and `engagement_analytics`.  I only say this because my gut feeling is that we're not going to want to hold onto the engagements forever (since the ideal situation is to have a lots and lots of engaged users) and their usefulness is very time limited.  However, the aggregation of engagements is generally useful longterm.

So if if engagements ( `{post id, pushToken, engagement, createdAt}` ) has a TTL on `createdAt` (bunfight to come later), engagement_analytics could be `{post id, count, first engagement timestamp, updatedAt}`

**Comment by rsinger** [comment]
@galenriley What model does `post` refer to?

**Comment by galenriley** [comment]
Thanks for your feedback and questions. Bringing your perspective into the project is super valuable.

---

Two collections make some sense, though I should point out that (so far) we are only considering tracking engagement with push notifications.
- not all posts have push notifications
- most engagement will be within a few seconds to a few hours after the post is published, so action on the API, and whatever it does, will be fleeting

Maybe we call the first collection notificationengagements, to not use up "engagements" on something that is more limited
---

postId is the _id of a document in feeditems

We've called the concept both, though post most often, most recently. The API is /feed/but everything in the client is a post. We could rename feeditems to posts, if you're down for that.

**Comment by galenriley** [comment]
Wow that's big I don't know why lol

**Comment by rsinger** [comment]
Thanks for the clarification!  That revises my thinking, then - we can almost certainly accomplish all of this simply with a single collection and the aggregation framework:
POST /notifications/{:post_id}/engagements -> `{ pushToken, timestamp}`
GET /notifications/{:post_id}/engagements -> raw dump of collection
GET /notifications/{:post_id} -> 
```
{
  id: , //mongoid
  count:,  //integer
  first_engagement:, //mongodate
  latest_engagement:, //mongodate
}
```

I'm not sold on the last route, yet, but that's easy enough to change once there's something to argue over.  I definitely think it's worth keeping the summary and the engagements separate, though, since the latter is an unbounded set.

NotificationEngagement can be as simple as:
```
{
  _id,
  postId,
  pushToken,
  engagementTimestamp,
  createdAt
}
```

**Comment by galenriley** [comment]
a note on the /notifications/ endpoint, for the sake of trivia

We named the current API /notifications/ arbitrairly, way back when the entire purpose of the app was to send songs to practice before the match. That functionality has been entirely supplanted by news feed, and we'll prolly be deprecating the API when we are sure everybody is updated.

soooooo you can have whatever in that space

**Comment by galenriley** [comment]
How do you feel bout going camelCase instead of underscores in your prop names?

**Comment by rsinger** [comment]
I am devoid of feelings

**Comment by galenriley** [comment]
I haven't had a sincere human emotion since 2018, so I'm with you on that one

**Comment by rsinger** [comment]
I started on this last night: https://github.com/rsinger/hooligan-hymnal-server/pull/1

@galenriley can you confirm a few assumptions?
1. A notification engagement can be defined as a push token and a feed item at a particular timestamp
2. There can only one instance of a specific pushtoken and feed item (i.e. the same push token and feed item can only ever appear in _one_ db document together)
3. There cannot be any auth/authz on the POST route, but we _can_ validate that the feed item and push token exist

**Comment by galenriley** [comment]
1. Yes sure
2. I'm going to hit the API when the user taps a push notification and, maybe, when the app is foregrounded and one is received (we can detect this case and process it like it was tapped). In the second case, the user might tap the notification a second time. Regardless, I think we are only interested in the first API hit and can ignore subsequent engagements. (@tacoman, thoughts?)
3. We haven't thought of a good way to ensure POST requests are valid for this feature, or the initial storing of push tokens and check-ins.

We expect the device's push token should already exist in the pushtokens collection and the post id already exists in feeditems.

**Comment by rsinger** [comment]
> We haven't thought of a good way to ensure POST requests are valid for this feature, or the initial storing of push tokens and check-ins

Realistically, if there is validation of the pushtoken and feed item _and_ we prevent the possibility of duplication, it's hard to see many opportunities for abuse outside of the usuals (e.g. DoS).

**Comment by galenriley** [comment]
The thing is- the only validation that the push token is correct is if it
matches a regex.

The bad news is, if one invalid token goes in the DB, then push
notifications break for everyone

**Comment by rsinger** [comment]
That sort of seems like a _different_ issue, though. If we're at point of notification engagement, that ship already sailed.

Although... can you explain a bit more how it breaks for everyone?

**Comment by galenriley** [comment]
You're totally right, re: ship sailing.

For the breaking- We ran into an issue way back before we understood anything that still occasionally plagues us to this day.

Long story short, Expo notes the bundle id and account when a pushToken is first generated/registered. Expo doesn't and won't tell us what that is. You send a notification alongside a list of all your push tokens. If expo notices that anything in the list doesn't match re: it's account/bundle id, then it doesn't send anything to anybody.

There are some ancient users out there who occasionally connect to the server without a recent update, the server registers their token with a different bundle id, and that spoils everything. I think we're confirming that a new property, .expoExperience, is with the push tokens before we save them (@tacoman either did or planned to add this), which should fix everything, but it's come up at least one more time since then.

tl;dr we've been traumatized by push tokens for a long, long time, and are gunshy when it comes to anything notification-related. Push didn't reliably work in Chattanooga until like November, even, and that's on an app that launched in preseason 2018.

**Comment by galenriley** [comment]
^ That's way too much exposition

**Comment by rsinger** [comment]
@galenriley Can we consider this closed?

**Comment by galenriley** [comment]
@tacoman is gonna test the notification engagement stuff on his dev server, then you'll get the satisfaction of closing the issue
